src/include/OSL/rendererservices.h

```diff
@@ -29,6 +30,125 @@ typedef const void* TransformationPtr;
 typedef void (*PrepareClosureFunc)(RendererServices*, int id, void* data);
 typedef void (*SetupClosureFunc)(RendererServices*, int id, void* data);
 
+enum class AttributeSpecBuiltinArg {
+    ShaderGlobalsPointer,  // void* (ideally ShaderGlobals*)
+    ShadeIndex,            // int
+    Derivatives,           // int (ideally bool)
+    Type,                  // long long (ideally TypeDesc)
+    ArrayIndex,            // int, Always zero for non-indexed array lookups.
+    ArrayLookup,           // int (ideally bool)
+    ObjectName,            // const char* (ideally ustring)
+    AttributeName,         // const char* (ideally ustring)
+};
+
+// TODO: replace with std::variant
+class AttributeSpecArg {
+    union {
+        AttributeSpecBuiltinArg m_builtin;
+        int32_t m_int32;
+        float m_float;
+        uint64_t m_uint64;
+        void* m_ptr;
+    };
+    uint32_t m_type;
```

Comment:
Perhaps a private enum to this class vs uint32_t for m_type,  not sure that really helps or not as its internal implementation detail.